util/src/main/java/com/google/cloud/hadoop/util/GoogleCredentialWithIamAccessToken.java

```diff
@@ -0,0 +1,73 @@
+package com.google.cloud.hadoop.util;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.api.client.auth.oauth2.TokenResponse;
+import com.google.api.client.googleapis.auth.oauth2.GoogleCredential;
+import com.google.api.client.http.HttpRequestInitializer;
+import com.google.api.client.http.HttpTransport;
+import com.google.api.client.json.jackson2.JacksonFactory;
+import com.google.api.client.util.Clock;
+import com.google.api.services.iamcredentials.v1.IAMCredentials;
+import com.google.api.services.iamcredentials.v1.model.GenerateAccessTokenRequest;
+import com.google.api.services.iamcredentials.v1.model.GenerateAccessTokenResponse;
+import java.io.IOException;
+import java.time.Instant;
+
+public class GoogleCredentialWithIamAccessToken extends GoogleCredential {
+  private final String serviceAccountName;
+  private final HttpRequestInitializer initializer;
+  private final HttpTransport transport;
+  private final Clock clock;
+
+  private static final String DEFAULT_ACCESS_TOKEN_LIFETIME = "3600s";
+  private static final String DEFAULT_SERVICE_ACCOUNT_NAME_PREFIX = "projects/-/serviceAccounts/";
+
+  public GoogleCredentialWithIamAccessToken(
+      String serviceAccountName,
+      HttpRequestInitializer initializer,
+      HttpTransport transport,
+      Clock clock)
+      throws IOException {
+    this.serviceAccountName = serviceAccountName;
+    this.initializer = initializer;
+    this.transport = transport;
+    this.clock = clock;
+    initialize();
+  }
+
+  private void initialize() throws IOException {
+    GenerateAccessTokenResponse accessTokenResponse = generateAccessToken();
+    Instant expirationTimeInInstant = Instant.parse(accessTokenResponse.getExpireTime());
+    setAccessToken(accessTokenResponse.getAccessToken());
+    setExpirationTimeMilliseconds(expirationTimeInInstant.toEpochMilli());
+  }
+
+  @Override
+  protected TokenResponse executeRefreshToken() throws IOException {
+    GenerateAccessTokenResponse accessTokenResponse = generateAccessToken();
+    Instant expirationTimeInInstant = Instant.parse(accessTokenResponse.getExpireTime());
+    Long expirationTimeMilliSeconds = expirationTimeInInstant.toEpochMilli();
+    return new TokenResponse()
+        .setAccessToken(accessTokenResponse.getAccessToken())
+        .setExpiresInSeconds(
+            expirationTimeMilliSeconds == null
```

Comment:
Do we want to create never-expiring tokens? Or if null is passed some default value is used?

Reply:
We use default value of 3600s for now, which is the longest the IAM access token is valid for. If for some reason the server didn't set the expirationValue the token will just be valid for whatever value set in request.

Reply:
Can it be `null` though? Seems like code will fail with `NullPointer` exception before reaching this point.

Reply:
It fails with NullPointerException if accessToken is null but not expiration time. I can see the same from TokenResponse class, access token is not allowed to be null but not expiration time, hence the logic.

Reply:
I meant that `Instant.parse` call will fail if expiration is null: https://docs.oracle.com/javase/8/docs/api/java/time/Instant.html#parse-java.lang.CharSequence-

So you either need to remove null check here all together, or move it before response  field parsing.

Reply:
Refactored to avoid this issue.